### ISSUE TITLE
Add one-electron transfer electrochemistry toy problem

### ISSUE THREAD

**Issue by rcw5890** [pints-team/pints#1039]
We are thinking about adding as a Pints toy problem the two species surface confined electrochemistry reaction.

A + e-  <->  B

The toy problem could use the dimensionless form of the equations with one-parameter capacitance, which are given below. Let _θ_ indicate the proportion of A, _i_ indicate the current, and _e_ indicate the input potential.

dθ/dt = k_0 ((1 - θ) exp((1 - α) (e_r - e_0)) - θ exp(-α (e_r - e_0)))
i = γ * de_r/dt + ζ * dθ/dt
e_r = e - ρ * i

The six dimensionless equation parameters are:
α = charge transfer coefficient
γ = capacitance
ρ = uncompensated resistance
k_0 = reaction kinetics parameter
e_0 = reversible potential
ζ = electrode coverage

_e_ gives the input potential. This could be fixed at a reasonable sinusoidal form.

Further details:
Adamson, Hope, et al. "Analysis of HypD disulfide redox chemistry via optimization of fourier transformed ac voltammetric data." _Analytical Chemistry_ 89.3 (2017): 1565-1573.

**Comment by martinjrobins** [comment]
sounds good, how were you thinking of doing the time integration? I don't think scipy has a dae solver

**Comment by rcw5890** [comment]
I was happy with the DAE solver in Julia, but at this point we presumably don't want to add Julia as a dependency for the toy problems. Is there a straight Python/scipy approach you would recommend?

**Comment by martinjrobins** [comment]
I havn't found anything suitably lightweight enough. Various packages wrap sundials (https://web.casadi.org/, https://scikits-odes.readthedocs.io/en/latest/). Casadi is pip installable, but quite big. Scikits odes requires you to download and compile sundials manually

**Comment by ben18785** [comment]
Just a quick one -- aren't we already using some sort of finite differences
scheme here to solve? Couldn't we just use that rather than rely on outside
packages...

**Comment by rcw5890** [comment]
Yes, I do have a finite difference implementation just in numpy which I was planning to start with. Of course, it's more lines of code/slower than what these libraries enable.

**Comment by martinjrobins** [comment]
I have c++ code to solve this in the electrochemistry repo, this is already wrapped in python. You could use this, but it would complicate the pints install process..

**Comment by ben18785** [comment]
If this is going to be a real pain, I'd say we could drop it (or make it an
example of how to interface with Julia). I just thought it might make a
good additional example (especially seeing as Dave likes this model).

**Comment by martinjrobins** [comment]
if you wanted to make an example, I would try CasADI, it's pretty good and can automatically generate jacobians, sensitivities etc. But I wouldn't have CasADI be a dependency for pints

**Comment by ben18785** [comment]
Yep, really like this idea! I'd like to see how it works for an example...

**Comment by MichaelClerx** [comment]
Seems too hard for PINTS proper. Can we make this a side project? Would agree with copying martins code and ideally wrapping something clever e.g. CVODE

**Comment by ben18785** [comment]
Let's chat about this tomorrow -- I think a notebook where we trial Casadi
would be really nice. But we wouldn't have Casadi as a dependency.

**Comment by MichaelClerx** [comment]
Let's!

That notebook would be untestable within this repo, so outside still seems better to me. We should probably have an "advanced" section in the examples anyway, where we refer to all the outside projects? (e.g. emupints, pints with tensorflow, etc.)
